core/src/main/java/io/grpc/internal/DelayedClientTransport2.java

```diff
@@ -105,54 +144,63 @@ public final Runnable start(Listener listener) {
+          if (lastPicker == null) {
+            return createPendingStream(method, headers, callOptions, statsTraceCtx);
+          }
+          picker = lastPicker;
+          pickerVersion = lastPickerVersion;
         }
-        picker = lastPicker;
       }
-    }
-    if (picker != null) {
-      while (true) {
-        PickResult pickResult = picker.pickSubchannel(callOptions.getAffinity(), headers);
-        ClientTransport transport = GrpcUtil.getTransportFromPickResult(
-            pickResult, callOptions.isWaitForReady());
-        if (transport != null) {
-          return transport.newStream(method, headers, callOptions, statsTraceCtx);
-        }
-        // This picker's conclusion is "buffer".  If there hasn't been a newer picker set
-        // (possible race with reprocess()), we will buffer it.  Otherwise, will try with the new
-        // picker.
-        synchronized (lock) {
-          if (shutdown) {
-            break;
+      if (picker != null) {
+        while (true) {
+          PickResult pickResult = picker.pickSubchannel(callOptions.getAffinity(), headers);
+          ClientTransport transport = GrpcUtil.getTransportFromPickResult(
+              pickResult, callOptions.isWaitForReady());
+          if (transport != null) {
+            return transport.newStream(method, headers, callOptions, statsTraceCtx);
           }
-          if (picker == lastPicker) {
-            return createPendingStream(method, headers, callOptions, statsTraceCtx);
+          // This picker's conclusion is "buffer".  If there hasn't been a newer picker set
+          // (possible race with reprocess()), we will buffer it.  Otherwise, will try with the new
+          // picker.
+          synchronized (lock) {
+            if (shutdown) {
+              break;
+            }
+            if (pickerVersion == lastPickerVersion) {
+              return createPendingStream(method, headers, callOptions, statsTraceCtx);
+            }
+            picker = lastPicker;
+            pickerVersion = lastPickerVersion;
           }
-          picker = lastPicker;
         }
       }
+      return new FailingClientStream(Status.UNAVAILABLE.withDescription(
+              "Channel has shutdown (reported by delayed transport)"));
+    } finally {
+      channelExecutor.drain();
     }
-    return new FailingClientStream(Status.UNAVAILABLE.withDescription(
-            "Channel has shutdown (reported by delayed transport)"));
   }
 
   @Override
   public final ClientStream newStream(MethodDescriptor<?, ?> method, Metadata headers) {
     return newStream(method, headers, CallOptions.DEFAULT, StatsTraceContext.NOOP);
   }
 
+  // Caller must call channelExecutor.drain() outside of lock because this method may schedule
+  // tasks on channelExecutor
   @GuardedBy("lock")
   private PendingStream createPendingStream(MethodDescriptor<?, ?> method, Metadata headers,
       CallOptions callOptions, StatsTraceContext statsTraceCtx) {
```

Comment:
I don't see it really helps. I tend to avoid passing member variable to a non-static method because it only increases confusion for me when reading the method body.